libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/extensions/WaypointEx.kt

```diff
@@ -0,0 +1,43 @@
+@file:JvmName("WaypointEx")
+
+package com.mapbox.navigation.base.internal.extensions
+
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.navigation.base.internal.route.Waypoint
+import com.mapbox.navigation.base.trip.model.RouteProgress
+
+/**
+ * Return waypoints that are requested explicitly
+ */
+fun List<Waypoint>.requestedWaypoints(): List<Waypoint> =
+    this.filter { it.type != Waypoint.EV_CHARGING }
+
+/**
+ * Return waypoints that tracking in [RouteProgress.currentLegProgress]#legIndex and based on
+ * [DirectionsRoute.legs] index
+ */
+fun List<Waypoint>.legsWaypoints(): List<Waypoint> =
+    this.filter { it.type != Waypoint.SILENT }
+
+/**
+ * Return the index of **next requested** coordinate.
+ *
+ * For instance, EV waypoints are not requested explicitly, so they are not taken into account.
+ */
+fun indexOfNextRequestedCoordinate(
+    waypoints: List<Waypoint>,
+    remainingWaypoints: Int,
+): Int? {
+    if (remainingWaypoints > waypoints.size) {
+        return null
+    }
+    val nextRequestedWaypoint = waypoints
+        .subList(waypoints.size - remainingWaypoints, waypoints.size)
+        .requestedWaypoints()
+        .firstOrNull()
+        ?: return null
+
+    return waypoints
+        .requestedWaypoints()
```

Comment:
logistic I think might have a lot of waypoints (probably put to 100, but I don't have any statistics). 
Do you see how can we get the index of the next requested waypoint another way?

Reply:
```
    var requestedIndex = 0
    waypoints.forEachIndexed { index, waypoint ->
        if (waypoint.type == Waypoint.REGULAR || waypoint.type == Waypoint.SILENT) {
            if (index >= waypoints.size - remainingWaypoints) {
                return requestedIndex
            }
            requestedIndex++
        }
    }
    return null
```

I think it would be equivalent.